src/Elastic.Apm.AspNetCore/DiagnosticListener/AspNetCoreDiagnosticListener.cs

```diff
@@ -35,14 +33,15 @@ public void OnNext(KeyValuePair<string, object> kv)
 
 			var exception = kv.Value.GetType().GetTypeInfo().GetDeclaredProperty("exception").GetValue(kv.Value) as Exception;
 
-			var transaction = _agent.Tracer.CurrentTransaction;
+			var transaction = (Transaction)_agent.Tracer.CurrentTransaction;
```

Comment:
Hmm.. this cast is a little bit unfortunate - with this it MUST be a `Transaction`, otherwise it will throw an exception - so e.g. we can't just replace to `NoOpTransaction` later if needed.

I guess the root cause of the challenge here is that we don't dependency inject the configuration anymore, instead it comes through the `Transaction`. I'm not sure this is good - could we maybe find another way to flow the `ConfigSnapshot`?

Reply:
It's not like it's the only place where we assume that `ITransaction` is actually `Transaction` - for example https://github.com/elastic/apm-agent-dotnet/blob/a73b10f0f6d61de45786f10501120644b9a82ff7/src/Elastic.Apm/Report/PayloadSenderV2.cs#L219)

If/when we need to have other implementations of `ITransaction` we will revisit it. For example we can introduce `TransactionBase` and move `ConfigSnapshot` there.

Reply:
I'd consider a pattern matching based on types very different from a hard cast.

But I think we are good here. In https://github.com/elastic/apm-agent-dotnet/pull/608 I moved this part out from here, so probably we can also get rid of the cast once we merge that. 

Context: In #608 I realized that capturing request body here is not a very good thing, because this is a sync method and reading the body synchronously can cause thread starvation. So I'd say let's leave it as is, and in #608 this will evolve.